### PR TITLE
convert the RadThermalWave derives to C++

### PR DESCRIPTION
**Pull request by zingale**
## PR checklist

- [ ] test suite needs to be run on this PR
- [ ] this PR will change answers in the test suite to more than roundoff level
- [ ] all newly-added functions have docstrings as per the coding conventions
- [ ] the `CHANGES` file has been updated, if appropriate
- [ ] if appropriate, this change is described in the docs

### CODE REVIEW

**Comment by zingale** [comment]
tests: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2020-09-27-001/index.html

File: Exec/radiation_tests/RadThermalWave/Problem_Derive.cpp [hunk 1]
```diff
@@ -4,56 +4,117 @@
...
+  auto const der = derfab.array();
+
+  amrex::ParallelFor(bx,
+  [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k) noexcept
+  {
+
+    Real loc[3];
+
+    loc[0] = problo[0] + (static_cast<Real>(i) + 0.5_rt) * dx[0];
+    loc[1] = problo[1] + (static_cast<Real>(j) + 0.5_rt) * dx[1];
```
Comment (maxpkatz) [inline Exec/radiation_tests/RadThermalWave/Problem_Derive.cpp:32]: I think we need to guard this against invalid accesses in 1D and 2D.

File: Exec/radiation_tests/RadThermalWave/Problem_Derive.cpp [hunk 2]
```diff
@@ -4,56 +4,117 @@
...
+                         std::lgamma(1.5_rt));
+
+    Real num = 3.0_rt * pe + 2.0_rt;
+    Real fac1 = std::pow(2.0_rt, pe - 1.0_rt) * pe * std::pow(M_PI, pe);
+    Real ex1 = 1.0_rt / (3.0_rt * pe + 2.0_rt);
+    Real fac2 = std::pow(pfac, pe / (3.0_rt * pe + 2.0_rt));
+
+    Real xi0 = std::pow(num / fac1, ex1) * fac2;
+
+    fac1 = a * std::pow(Q, pe) * max(time, 1.e-50_rt);
```
Comment (maxpkatz) [inline Exec/radiation_tests/RadThermalWave/Problem_Derive.cpp:53]: amrex::max

File: Exec/radiation_tests/RadThermalWave/Problem_Derive.cpp (hunk 1 above)
Comment (zingale) [reply Exec/radiation_tests/RadThermalWave/Problem_Derive.cpp:33]: fixed (in 2 places)

File: Exec/radiation_tests/RadThermalWave/Problem_Derive.cpp (hunk 2 above)
Comment (zingale) [reply Exec/radiation_tests/RadThermalWave/Problem_Derive.cpp]: fixed

**Comment by zingale** [comment]
updated tests: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2020-09-27-003/index.html

**Review by maxpkatz — APPROVED** [review]

**Pull request merged**